### DISCUSSION TITLE
User data safe getter and setter

### DISCUSSION THREAD

**Discussion by DBLouis** [fltk-rs/fltk-rs#1450] [Ideas]
Hello
I see that there are unsafe internal functions to use user data associated to a widget. Do you think it would be possible to expose safe wrappers to those. Using std Any for example. I can make a PR maybe.

**Comment by MoAlyousef** [comment]
Hello
fltk-rs uses the user_data as the function object to be called with a callback. So I don’t think a setter can be ensured to be safe, especially when the set_callback function is set as well. 
I’m not sure though, so if you have a proof of concept I would like to have a look.

**Reply by DBLouis** [reply]
What if it is simply documented that you can only use one or the other, and the functions returns options accordingly?

**Reply by MoAlyousef** [reply]
This could be possible. A helper function could also add a callback to the widget which doesn’t require the use of the user_data

**Reply by MoAlyousef** [reply]
It seems passing through a void pointer makes std::any::Any lose its type info:
```rust
use std::any::Any;
use std::os::raw::c_void;

fn do1(r: *mut dyn Any) {
    let a: Box<dyn Any> = unsafe { Box::from_raw(r) };
    let a: i32 = *a.downcast().unwrap();
    println!("{:?}", a);
}

fn do2(r: *mut c_void) {
    let a: Box<dyn Any> = unsafe { Box::from_raw(r) };
    let a: i32 = *a.downcast().unwrap();
    println!("{:?}", a);
}

fn main() {
    let a: Box<dyn Any> = Box::new(1_i32);
    do1(Box::into_raw(a) as _); // works
    do2(Box::into_raw(a) as _); // doesn't work
}
```

**Reply by DBLouis** [reply]
Yes it's tricky. I remember a similar challenge. I think you need the ptr_metatada unstable feature to make it work.
https://doc.rust-lang.org/stable/std/ptr/fn.from_raw_parts.html
